### PR TITLE
Add go profile instrumentation to kubectl

### PR DESCRIPTION
**Pull request by dlespiau**
This commit adds two new global options to kubectl: --profile and
--profile-output, writing out go profiles to disk to debug interesting and
unexpected kubectl behaviour.

As an example, here is how to capture a block file, eg. for how long are we
blocked on I/O and where?

$ kubectl get nodes --profile=block -v6
$ go tool pprof -png ./profile.pprof > out.png
$ google-chrome out.png

**What this PR does / why we need it**:

This instrumentation is useful to debug abnormal kubectl behaviour or do performance work.

Fixes: #68679

```release-note
kubectl has gained new --profile and --profile-output options to output go profiles
```

As an example here's a block profile for the `kubectl get nodes` command above, rebuilding its cache. The api server isn't local.

screenshot from 2018-09-14 16-29-33

### CODE REVIEW

**Comment by neolit123** [comment]
/kind feature
up to the maintainers.

**Comment by dlespiau** [comment]
/retest

The pull-kubernetes-e2e-kops-aws failure seems like a flake.
The pull-kubernetes-verify test is still using the previous version of the PR:

```
I0914 16:31:40.316] Args: --job=pull-kubernetes-verify --service-account=/etc/service-account/service-account.json --upload=gs://kubernetes-jenkins/logs --job=pull-kubernetes-verify --repo=k8s.io/kubernetes=master:94e59f1636a8f8b1566ef35180519d4de5586b79,68681:8653e09382d6becd8a72d0bc06a47e7020008a2b --service-account=/etc/service-account/service-account.json --upload=gs://kubernetes-jenkins/pr-logs --timeout=75 --scenario=kubernetes_verify -- '--branch=${PULL_BASE_REF}' --script=./hack/jenkins/verify-dockerized.sh
```

8653e09382d6becd8a72d0bc06a47e7020008a2b is the previous commit. Hoping /retest will do the right thing.

File: pkg/kubectl/cmd/profiling.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	// Block and mutex profiles need a call to Set{Block,Mutex}ProfileRate to
+	// output anything. We choose to sample all events.
+	case "block":
+		runtime.SetBlockProfileRate(1)
+		return nil
+	case "mutex":
+		runtime.SetMutexProfileFraction(1)
+		return nil
+	default:
+		if profile := pprof.Lookup(profileName); profile == nil {
```
Comment (apelisse) [inline pkg/kubectl/cmd/profiling.go]: What's the goal of this profile? Does it do something?

File: pkg/kubectl/cmd/profiling.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	switch profileName {
+	case "":
+		return nil
+	case "cpu":
+		pprof.StopCPUProfile()
+	case "heap":
+		runtime.GC()
+		fallthrough
+	default:
+		profile := pprof.Lookup(profileName)
```
Comment (apelisse) [inline pkg/kubectl/cmd/profiling.go]: I wasn't sure what `profile` was used for in the init function, but now it makes sense, great.

**Review by apelisse — COMMENTED** [review]
Looks good to me, it should either fail or default if no profile is specified though

File: pkg/kubectl/cmd/profiling.go (hunk 1 above)
Comment (dlespiau) [reply pkg/kubectl/cmd/profiling.go]: This is just a map lookup to check if profileName is the name of an existing profile, validating the input parameter.

Given it prompted a question, it's definitely worth a comment, added it!

**Comment by dlespiau** [comment]
If no profile is specified, `profileName` is `""` and those functions do nothing, which is the intended behaviour. If `profileName` is non zero, the profile name is validated by initProfiling to make sure it's the name of a valid profile and fails if it isn't. I think we're covered then.

**Comment by apelisse** [comment]
> If no profile is specified, profileName is "" and those functions do nothing

Yeah, I don't think I agree. Running `kubectl --profile whatever` shouldn't skip profiling, that's too confusing.

**Comment by dlespiau** [comment]
Ah I see what you mean, I think that's a general problem with options though :)

```
$ kubectl --kubeconfig get nodes
Error: unknown command "nodes" for "kubectl"
Run 'kubectl --help' for usage.
unknown command "nodes" for "kubectl"
```

`--profile` alone has the same behaviour:

```
$ ./_output/bin/kubectl --profile get nodes
Error: unknown command "nodes" for "kubectl"
Run 'kubectl --help' for usage.
unknown command "nodes" for "kubectl"
```

`--profile` alone isn't meaningful, just like `--kubeconfig` alone isn't. The parser takes the next token as  the `--profile` option and tries to find the `nodes` sub-command, which doesn't exist. I tried to play with `--` to terminate the option list and force an empty `--profile` option but to no avail. `--profile` on its own is simply not valid, just like `--kubeconfig` on its own.

To achieve meaningful `--profile` option, we'd need to have it a boolean option I believe and the UX would become:
```
$ kubectl --profile --profile-kind=block --profile-output=foo.pprof get nodes
```
with `--profile-kind` defaulting to, say, `cpu`. I have a preference for only having two options.

Just in case the `whatever` in your example meant a bogus profile name,  the current code errors out:
```
$ ./_output/bin/kubectl --profile whatever get nodes
[...]
unknown profile 'whatever'
```

**Comment by apelisse** [comment]
Sorry, I made a mistake ... I meant `$ kubectl --profile= get pods`. It's still weird if this does no profiling.

**Comment by apelisse** [comment]
> Sorry, I made a mistake ... I meant $ kubectl --profile= get pods. It's still weird if this does no profiling.

or maybe it's not, maybe you should have an option named "none", and make it the defautl?

**Comment by dlespiau** [comment]
Oh I like the `"none"` idea! Thanks for taking the time to think about it and make a great suggestion!

Now, giving a `""` string errors out as such:
```
./_output/bin/kubectl --profile= get nodes
[...]
unknown profile ''
```

**Comment by dlespiau** [comment]
Thanks a lot for the review!

**Comment by seans3** [comment]
Would it be possible to add a unit test (e.g. profile_test.go)?

**Comment by dlespiau** [comment]
Thinking a bit about it, we could maybe add tests along the lines of:
  - Check that giving a bogus profile name leads initProfiling to return an error
  - Check that giving a valid profile name generates a file (and maybe even check that is indeed a pprof file). Maybe cycle through all profile names to check all well-known names generate profiles.

**Pull request merged**